bika/lims/browser/fields/historyawarereferencefield.py

```diff
@@ -118,84 +107,51 @@ def set(self, instance, value, **kwargs):
             #e.g. if i want to store the reference UIDs into an SQL field
             ObjectField.set(self, instance, self.getRaw(instance), **kwargs)
 
-    security.declarePrivate('get')
-
-    def get(self, instance, aslist=False, **kwargs):
-        """get() returns the list of objects referenced under the relationship.
+    def get_referenced_version(self, instance, reference):
+        """Returns the object from the reference history that matches with the
+        version the instance points to
         """
-        try:
-            uc = getToolByName(instance, "uid_catalog")
-        except AttributeError as err:
-            logger.error("AttributeError: {0}".format(err))
-            return []
-
-        try:
-            res = instance.getRefs(relationship=self.relationship)
-        except:
-            pass
-
+        # Version of the referenced object to which the instance points to.
+        # If no version found or None, assume the instance points to the first
+        # version created (otherwise, it should have a value)
+        referenced_version = getattr(instance, "reference_versions", {})
+        referenced_version = referenced_version.get(reference.UID(), None) or 0
+
+        # Current version of the referenced object.
+        # If the object has not yet a version explicitly set (no history yet),
+        # assume this is the first version created
+        reference_version = getattr(reference, "version_id", None) or 0
+        if reference_version == referenced_version:
+            # The instance points to the latest version, no need to look for
+            # previous versions
+            return reference
+
+        # The instance points to a previous version, need to get the exact
+        # previous version the instance points to
         pr = getToolByName(instance, 'portal_repository')
+        version_data = pr._retrieve(reference, selector=referenced_version,
+                                    preserve=(), countPurged=True)
```

Comment:
I mean in case of `None`, why not simply retrieve the original object?

Reply:
When `selector` param is `None`, `pr._retrieve` always returns the last/current version (instead of the original one). `_retrieve` delegates the action to `retrieve` func from `ArchivistTool` (note the comment in there):

https://github.com/plone/Products.CMFEditions/blob/68fdce45dcac1ee921c9f4f9d78dbe8f84eaddee/Products/CMFEditions/ArchivistTool.py#L332-L346

And for `None`, `LazyHistory` always return the last/most recent version of the history. Makes sense, but in our case, must be the other way round.

Reply:
I thought more about bypassing the whole archivist when we're in the initial version and just return the original object. Is that possible?